proposals/0177-program-runtime-abiv2.md

```diff
@@ -0,0 +1,289 @@
+  - Lamports: `u64`
+  - Account payload: `&[u8]` which is composed of:
+    - Pointer to account payload: `u64`
+    - Account payload length: `u64`
+
+#### Instruction area
+
+For each transction, the program runtime must also preapre two memory regions. 
+The first one is a readonly region starting at `0x600000000`. It must be 
+updated at each CPI call edge. The contents of this region are the following:
+
+- For each instruction in transaction:
+  - Index in transaction of program account to be executed: `u64`
+  - CPI nesting level: `u16`
+  - Index of parent instruction (`u16::MAX` for top-level instructions): `u16`
+  - Reference to a slice of instruction accounts `&[InstructionAccount]`, 
+    consisting of:
+    - Pointer to beginning of slice: `u64` (points to the `0x700000000` 
+      region)
+    - Number of elements in slice: `u64`
+  - Instruction data `&[u8]`, which is composed of:
+    - Pointer to data: `u64`
+    - Length of data: `u64`
+
+Let `InstrucionAccount` contain the following fields:
+
+  - Index to transaction account: `u16`
+  - Signer flag: `u8` (1 for signer, 0 for non-singer)
+  - Writable flag: `u8` (1 for writable, 0 for readonly)
+
+The second region is readonly and starts at `0x700000000`. It must also be 
+updated at each CPI call edge. This region contains an array of all 
+`InstructionAccounts` for every instruction in the transction. In other words:
+
+- For instruction in transaction:
+  - For each account in instruction:
+    - `InstructionAccount`, consisting of:
+      - Index to transaction account: `u16`
+        - Signer flag: `u8` (1 for signer, 0 for non-singer)
+        - Writable flag: `u8` (1 for writable, 0 for readonly)
+
+#### Return data scratchpad
+
+A writable memory region starting at `0x800000000` must be mapped in for the
+return-data scratchpad.
+
+### Accounts area
+
+For each unique (meaning deduplicated) instruction account the payload must
+be mapped in at `0x900000000` plus `0x100000000` times the index of the
+**transaction** account (not the index of the instruction account). Only if the
+instruction account has the writable flag set and is owned by the current
+program it is mapped in as a writable region. The writability of a region must
+be updated as programs through out the transaction modify the account metadata.
+
+The runtime must only map the payload for accounts that belong in the current
+executing instruction. The payload for accounts belonging to sibling instructions
+must NOT be mapped.
+
+### Instruction payload area
+
+For each instruction, the runtime must map its payload at address 
+`0x10900000000` plus `0x100000000` times the index of the instruction in the 
+trasaction. All instruction payload mappings are readonly.
+
+One extra writable mapping must be created after the last instruction payload 
+area to be the CPI scratch pad, i.e. at address `0x10900000000` plus 
+`0x100000000` times the number of instructions in the transaction. Its purpose 
```

Comment:
Can this be a static location? Removes a math operation to get the  cpi scratchpad ptr